src/backend/distributed/operations/worker_node_manager.c

```diff
@@ -418,19 +418,21 @@ NodeIsPrimaryWorker(WorkerNode *node)
 
 /*
  * CoordinatorAddedAsWorkerNode returns true if coordinator is added to the
- * pg_dist_node. This function also acquires ShareLock on pg_dist_node
- * and does not release it to ensure that existency of the coordinator in
- * metadata won't be changed until the end of transaction.
+ * pg_dist_node. This function also acquires AccessShareLock on pg_dist_node
+ * and does not immediately releases it to prevent any self deadlocks
+ * on operations that acquire heavier locks on pg_dist_node.
  */
 bool
 CoordinatorAddedAsWorkerNode()
 {
 	bool groupContainsNodes = false;
 
-	LockRelationOid(DistNodeRelationId(), ShareLock);
+	LockRelationOid(DistNodeRelationId(), AccessShareLock);
```

Comment:
we no longer guarantee that the coordinator is not removed with an AccessShareLock.

it also seems a bit odd that this function concerns itself with that at all.

maybe we just remove the lock?

Reply:
makes sense